tensorflow_datasets/core/custom_dataset/image_folder.py

```diff
@@ -0,0 +1,181 @@
+    root_dir = os.path.expanduser(root_dir)
+    super(ImageFolder, self).__init__(
+        data_dir=root_dir, version=str(self.VERSION), config=None)
+
+    # Reset `_data_dir` as it should not change to DATA_DIR/Version
+    self._data_dir = root_dir
+
+    # dict[split_name][label_name] = list(img_paths)
+    self._split_label_images, labels = _get_split_label_images(root_dir)
+
+    # Calculate and update DatasetInfo labels
+    self.info.features['label'].names = labels
+
+    # Calculate and update DatasetInfo splits
+    splits = {
+        split_name: sum(len(l)
+                        for l in self._split_label_images[split_name].values())
+        for split_name in self._split_label_images
+    }
+    split_dict = core.SplitDict(self.name)
+    for split_name, size in splits.items():
+      split_dict.add(
+          core.SplitInfo(name=split_name, shard_lengths=[size]))
+    self.info.update_splits_if_different(split_dict)
+
+  def _info(self):
+    return core.DatasetInfo(
+        builder=self,
+        description='Generic image classification dataset.',
+        features=core.features.FeaturesDict({
+            'image': core.features.Image(),
+            'label': core.features.ClassLabel(),
+            'image/filename': core.features.Text(),
+        }),
+        supervised_keys=('image', 'label'),
+    )
+
+  def _download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def _map_fn(self, path):
+    image_path = tf.strings.split(path, os.path.sep)[-3:]
+    img = tf.io.read_file(path)
+    img = tf.image.decode_jpeg(img, channels=3)
+    img = tf.image.convert_image_dtype(img, tf.uint8)
+    label = image_path[1].numpy()
+    return [
+        img,
+        self.info.features['label'].str2int(label),
+        tf.strings.reduce_join(image_path)
+    ]
+
+  def _process_ds(self, path):
+    output = tf.py_function(self._map_fn,
+                            inp=[path],
+                            Tout=list(self.info.features.dtype.values()))
+    output = {
+        'image': output[0],
+        'label': output[1],
+        'image/filename': output[2],
+    }
+    for key, value in output.items():
+      value.set_shape(self.info.features.shape[key])
+    return output
+
+  def _as_dataset(self, split, shuffle_files=False, **kwargs):
+    """Generate dataset for given split"""
+    del kwargs
+    if split not in self.info.splits.keys():
+      raise ValueError('Split name {} not present in {}'
+                       .format(split, self._split_label_images.keys()))
+
+    imgs = sorted(
+        img for l in self._split_label_images[split].values() for img in l)
+
+    ds = tf.data.Dataset.list_files(imgs)
+    ds = ds.map(self._process_ds,
+                num_parallel_calls=tf.data.experimental.AUTOTUNE)
+    if shuffle_files:
```

Comment:
The shuffling should happen before the images are loaded in memory, otherwise this will load the full dataset in memory.